libs/state/src/lib/cell-summary/cell-summary.state.ts

```diff
@@ -0,0 +1,91 @@
+import { Injectable } from '@angular/core';
+import { Action, State, StateContext } from '@ngxs/store';
+import { ComputeAggregate } from './cell-summary.actions';
+import { Aggregate, AggregateRow, Cell, CellSummary, CellSummaryStateModel } from './cell-summary.model';
+
+/** State handling cell summary data */
+@State<CellSummaryStateModel>({
+  name: 'cellsummary',
+  defaults: {
+    summaries: {},
+    aggregate: {},
+  },
+})
+@Injectable()
+export class CellSummaryState {
+  /**
+   * computes aggregate data and stores in the current state
+   */
+  @Action(ComputeAggregate)
+  computeAggregate(ctx: StateContext<CellSummaryStateModel>, action: ComputeAggregate) {
+    const summaries = action.summaries;
+    const aggregateData: Aggregate = {};
+
+    for (const key in summaries) {
+      const label = summaries[key].label;
+      const entries = summaries[key].entries;
+
+      const counts = this.getCellCountMap(entries);
+
+      const [rows, columns] = this.getRowsAndColumns(entries, counts);
+
+      aggregateData[key] = {
+        label: label,
+        columns: columns as string[],
+        rows: rows as AggregateRow[],
+      };
+    }
+
+    ctx.patchState({
+      aggregate: aggregateData,
+    });
+  }
+
+  /** Gets a count of the number of occurrences of each unique cellId in an array of entries,
+   * and returns an object mapping each cellId to its corresponding count. */
+  private getCellCountMap(entries: CellSummary[string]['entries']) {
+    const counts: { [key: string]: number } = {};
+
+    entries.forEach((entry) => {
+      const cellId = entry.cell.id;
+
+      if (cellId in counts) {
+        counts[cellId] += entry.count;
+      } else {
+        counts[cellId] = entry.count;
+      }
+    });
+
+    return counts;
+  }
+
+  /** Returns an array of rows and columns based on an array of entries */
+  getRowsAndColumns(entries: CellSummary[string]['entries'], counts: { [key: string]: number }) {
```

Comment:
I don't think the logic here gives the correct result. Let's go over it together